packages/components/form/src/form.vue

```diff
@@ -1,258 +1,158 @@
+  if (!props.model) {
+    debugWarn(COMPONENT_NAME, 'model is required for resetFields to work.')
+    return
   }
+  filterFields(fields, properties).forEach((field) => field.resetField())
 }
 
-export interface Callback {
-  (isValid?: boolean, invalidFields?: ValidateFieldsError): void
+const clearValidate: FormContext['clearValidate'] = (props = []) => {
+  filterFields(fields, props).forEach((field) => field.clearValidate())
 }
 
-export default defineComponent({
-  name: 'ElForm',
-  props: {
-    model: Object,
-    rules: Object as PropType<FormRulesMap>,
-    labelPosition: String,
-    labelWidth: {
-      type: [String, Number],
-      default: '',
-    },
-    labelSuffix: {
-      type: String,
-      default: '',
-    },
-    inline: Boolean,
-    inlineMessage: Boolean,
-    statusIcon: Boolean,
-    showMessage: {
-      type: Boolean,
-      default: true,
-    },
-    size: String as PropType<ComponentSize>,
-    disabled: Boolean,
-    validateOnRuleChange: {
-      type: Boolean,
-      default: true,
-    },
-    hideRequiredAsterisk: {
-      type: Boolean,
-      default: false,
-    },
-    scrollToError: Boolean,
-  },
-  emits: ['validate'],
-  setup(props, { emit }) {
-    const fields: FormItemCtx[] = []
-
-    watch(
-      () => props.rules,
-      () => {
-        fields.forEach((field) => {
-          field.evaluateValidationEnabled()
-        })
+const validate = async (callback?: FormValidateCallback): Promise<void> =>
+  validateField(undefined, callback)
+
+const validateField: FormContext['validateField'] = async (
+  properties = [],
+  callback
+) => {
+  if (callback) {
+    validate()
+      .then(() => callback(true))
+      .catch((fields: ValidateFieldsError) => callback(false, fields))
+    return
+  }
 
-        if (props.validateOnRuleChange) {
-          validate(() => ({}))
-        }
-      }
-    )
+  const { model, scrollToError } = props
 
-    const formSize = useSize()
-    const prefix = 'el-form'
-    const formKls = computed(() => {
-      const { labelPosition, inline } = props
-      return [
-        prefix,
-        `${prefix}--${formSize.value}`,
-        labelPosition ? `${prefix}--label-${labelPosition}` : '',
-        inline ? `${prefix}--inline` : '',
-      ]
-    })
+  if (!model) {
+    debugWarn(COMPONENT_NAME, 'model is required for validate to work!')
+    return
+  }
+  if (fields.length === 0) {
+    return
+  }
 
-    const addField = (field: FormItemCtx) => {
-      if (field) {
-        fields.push(field)
-      }
-    }
+  const fds = filterFields(fields, properties)
+  if (!fds.length) {
+    debugWarn(COMPONENT_NAME, 'please pass correct props!')
+    return
+  }
 
-    const removeField = (field: FormItemCtx) => {
-      if (field.prop) {
-        fields.splice(fields.indexOf(field), 1)
-      }
-    }
+  let valid = true
+  let invalidFields: ValidateFieldsError = {}
+  let firstInvalidFields: ValidateFieldsError | undefined = undefined
```

Comment:
**Nitpick:** I do not think you need to explicitly `undefined` a variable.

Reply:
<img width="621" alt="image" src="https://user-images.githubusercontent.com/6481596/156924420-9d5b2a30-aad5-4cf0-bef1-152c51a82262.png">

Reply:
I was talking about doing:

```ts
let firstInvalidFields: ValidateFieldsError | undefined // no need do assignment like firstInvalidFields = undefined

```

Reply:
If you remove the `undefined` constrain meaning that your variable will always be instantiated but it wasn't in your case, that is why the type checker is complaining about it was been used before initiated.